### PR TITLE
Fix serialization and caching

### PR DESCRIPTION
**Pull request by AdrianSosic**
Due to continuing serialization problems that were thought to be related with caching, #277 deactivated core test caching and #294 was prepared to do the same for the full test environment.

This PR reactivates caching and instead refactors the class layout of `SKLearnClusteringRecommender` in an attempt to fix the root cause. Mysteriously, the top-level import of `sklearn.mixture.GaussianMixture` seems to cause trouble. While the reason is still unclear, turning it into a lazy import (which will also become handy later when making `scikit-learn` an optional dependency) seems to resolve the problem.

On a side note: deactivating slots for the recommenders solves the problem as well, which suggests that the root cause could be related to classes not being properly garbage collected (since `attrs` needs to create new classes when slots are activated), which could also explain that `GaussianMixtureClusteringRecommender` seemed to have improperly overridden methods after deserialization (for example, the  `__repr__` of a created Gaussian mixture recommender correctly pointed to its own class before serialization but to the `__repr__` of `SKLearnClusteringRecommender` after serialization – but weirdly only when executed in `tox`).

### CODE REVIEW

File: .github/workflows/regular.yml [hunk 1]
```diff
@@ -1,7 +1,8 @@
 # NOTES:
 # - The map syntax used for matrix is flagged red but actually works
 # - This runs everything in Python 3.10, 3.11 and 3.12
-# - No environments are cached due to space limit
+# - Environments are **not** cached in order to perform a full end-to-end test
```
Comment (AVHopp) [inline .github/workflows/regular.yml:4]: I'm confused, are environments cached ot not? The action looks like we cache stuff, can you elaborate?

File: baybe/recommenders/pure/nonpredictive/clustering.py [hunk 2]
```diff
@@ -54,6 +43,11 @@ class SKLearnClusteringRecommender(NonPredictiveRecommender, ABC):
     """Optional model parameter that will be passed to the surrogate constructor.
     This is initialized with reasonable default values for the derived child classes."""
 
+    @staticmethod
+    @abstractmethod
+    def _get_model_cls() -> type[ClusterMixin]:
```
Comment (AVHopp) [inline baybe/recommenders/pure/nonpredictive/clustering.py:48]: Why is this private? The previous class variable was public, so why the change?

**Review by AVHopp — APPROVED** [review]
Just two minor question. Approve for unblocking since I will be quite busy.

File: .github/workflows/regular.yml (hunk 1 above)
Comment (Scienfitz) [reply .github/workflows/regular.yml:4]: yep, we dont need caching here in regular

**Review by Scienfitz — APPROVED** [review]

File: .github/workflows/regular.yml (hunk 1 above)
Comment (AdrianSosic) [reply .github/workflows/regular.yml:4]: Thanks, good catch. I reverted @Scienfitz commit but didn't pay enough attention (it also reactivated caching in `regular.yml`, which I did not intend). Is now fixed 👍🏼

File: baybe/recommenders/pure/nonpredictive/clustering.py (hunk 2 above)
Comment (AdrianSosic) [reply baybe/recommenders/pure/nonpredictive/clustering.py:48]: Yes, in general I'd like to be a bit more cautious what we expose as public and what not. In the past, pretty much everything was public (like stuff in this module, which was written long time ago), even though not really needed. That only causes pain with backward compatibility when you want to refactor stuff. Since the method is really not expected to be used outside, it think its better to keep it private for now. We can always change when needed: making private stuff public is easy, the other way round not. So I took this opportunity....
